File: LeetCodeKotlin/BinaryTreeLevelOrderTraversal_102.kt
class BinaryTreeLevelOrderTraversal_102 {
    fun levelOrder(root: TreeNode?): List<List<Int>> {
        val result = mutableListOf<List<Int>>()
        if (root == null) return result

        val queue: ArrayDeque<TreeNode> = ArrayDeque()
        queue.add(root)

        while (queue.isNotEmpty()) {
            val levelSize = queue.size
            val currentLevel = mutableListOf<Int>()

            repeat(levelSize) {
                val node = queue.removeFirst()
                currentLevel.add(node.value)

                node.left?.let { queue.add(it) }
                node.right?.let { queue.add(it) }
            }
            result.add(currentLevel)
        }

        return result
    }
}

File: LeetCodeCpp/problems/trees/BinaryTreeLevelOrderTraversal_102.cpp
#include "BinaryTreeLevelOrderTraversal_102.h"

#include <queue>

using namespace std;

vector<vector<int>> BinaryTreeLevelOrderTraversal_102::levelOrder(
    TreeNode<int>* root) {
    vector<vector<int>> result;

    if (root == nullptr) {
        return result;
    }

    queue<TreeNode<int>*> nodes;
    nodes.push(root);

    while (!nodes.empty()) {
        const size_t levelSize = nodes.size();

        vector<int> level;
        level.reserve(levelSize);

        for (size_t i = 0; i < levelSize; ++i) {
            TreeNode<int>* const node = nodes.front();
            nodes.pop();

            level.push_back(node->val);

            if (node->left != nullptr) {
                nodes.push(node->left);
            }
            if (node->right != nullptr) {
                nodes.push(node->right);
            }
        }

        result.push_back(move(level));
    }

    return result;
}
